src/main/java/com/aws/greengrass/deployment/ConnectivityValidator.java

```diff
@@ -0,0 +1,483 @@
+import static com.aws.greengrass.mqttclient.MqttClient.MQTT_SOCKET_TIMEOUT_KEY;
+
+public class ConnectivityValidator implements AutoCloseable {
+    private static final Logger logger = LogManager.getLogger(ConnectivityValidator.class);
+
+    private final DeviceConfiguration deviceConfiguration;
+    private final SecurityService securityService;
+
+    private AwsIotMqttConnectionBuilder builder;
+    private MqttClientConnection connection;
+    private GreengrassV2DataClient greengrassV2DataClient;
+    private TlsContextOptions proxyTlsOptions;
+    private ClientTlsContext tlsContext;
+
+    @Inject
+    public ConnectivityValidator(DeviceConfiguration deviceConfiguration, SecurityService securityService) {
+        this.deviceConfiguration = deviceConfiguration;
+        this.securityService = securityService;
+    }
+
+
+    /**
+     * Validate Mqtt and HTTP connectivity using the nucleus config.
+     *
+     * @param totallyCompleteFuture future deployment result
+     * @param nucleusConfig         nucleus config to validate
+     * @param deployment            deployment
+     */
+    public boolean validateConnectivity(CompletableFuture<DeploymentResult> totallyCompleteFuture,
+                                        Map<String, Object> nucleusConfig, Deployment deployment) {
+        Integer timeoutSec;
+        timeoutSec = deployment.getDeploymentDocumentObj().getConfigurationValidationPolicy().timeoutInSeconds();
+        if (nucleusConfig != null) {
+            long configTimeout = Duration.ofSeconds(DEFAULT_TIMEOUT_SECOND).toMillis();
+            if (timeoutSec != null) {
+                configTimeout = Duration.ofSeconds(timeoutSec).toMillis();
+            }
+
+            if (configTimeout == 0 || !deviceConfiguration.isDeviceConfiguredToTalkToCloud()) {
+                logger.atDebug().log("Skipping connectivity validation");
+                return true;
+            }
+
+            // MQTT connectivity validation
+            boolean reconnected = false;
+            try {
+                logger.atDebug().log("Checking MQTT client can connect");
+                createMqttConnectionBuilder(nucleusConfig);
+                checkMqttConnection(configTimeout);
+                reconnected = true;
+            } catch (MqttConnectionProviderException e) {
+                logger.atError().cause(e).log("Failed to creat MQTT connection builder for connectivity validation");
+                totallyCompleteFuture
+                        .complete(new DeploymentResult(DeploymentResult.DeploymentStatus.FAILED_NO_STATE_CHANGE, e));
+            } catch (ComponentConfigurationValidationException e) {
+                logger.atError().cause(e).log("MQTT connectivity validation failed");
+                totallyCompleteFuture
+                        .complete(new DeploymentResult(DeploymentResult.DeploymentStatus.FAILED_NO_STATE_CHANGE, e));
+            } catch (InterruptedException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

**Remember**: Do this only for InterruptedException. Doing this for other types of exceptions may give rise to subtle bugs.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)